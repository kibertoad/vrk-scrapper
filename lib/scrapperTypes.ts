export type CandidateVotes = {
  name: string
  party: string
  districtVotes: number
  mailVotes: number
  totalVotes: number
  percentFromValid: number
  percentFromAll: number
}

export type DistrictResults = {
  attrHeaders: readonly string[]
  candidateValues: CandidateVotes[]
}
