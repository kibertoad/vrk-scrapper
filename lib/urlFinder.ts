// const urlRegex = /\'?srcUrl=\'(.*)\'.html\'/

export function findAllUrls(text: string): readonly string[] {
  const result = text.match(/\?srcUrl(.*)\.html/gi) ?? []
  return result.map((srcUrl) => {
    return srcUrl.replace('?srcUrl=', '')
  })
}
