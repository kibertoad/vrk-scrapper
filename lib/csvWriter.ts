import { createObjectCsvWriter } from 'csv-writer'
import { CandidateVotes } from './scrapperTypes'
import { stat, mkdir } from 'fs/promises'

function ensureDirectoryExists(dir: string) {
  return stat(dir).catch(() => mkdir(dir, { recursive: true }))
}

export async function writeSingleMandateResults(
  districtName: string,
  attrHeaders: readonly string[],
  candidateData: CandidateVotes[]
) {
  const dir = 'singleMandateByDistrict'
  await ensureDirectoryExists(dir)

  const headers = [
    {
      id: 'name',
      title: attrHeaders[0].toUpperCase(),
    },
    {
      id: 'party',
      title: attrHeaders[1].toUpperCase(),
    },
    {
      id: 'districtVotes',
      title: attrHeaders[2].toUpperCase(),
    },
    {
      id: 'mailVotes',
      title: attrHeaders[3].toUpperCase(),
    },
    {
      id: 'totalVotes',
      title: attrHeaders[4].toUpperCase(),
    },
    {
      id: 'percentFromValid',
      title: attrHeaders[5].toUpperCase(),
    },
    {
      id: 'percentFromAll',
      title: attrHeaders[6].toUpperCase(),
    },
  ]

  const csvWriter = createObjectCsvWriter({
    path: `${dir}/${districtName}.csv`,
    header: headers,
  })

  await csvWriter.writeRecords(candidateData)
}
