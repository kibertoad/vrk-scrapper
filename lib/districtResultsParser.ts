import parse5, { DefaultTreeParentNode, Document } from 'parse5'
import { CandidateVotes, DistrictResults } from './scrapperTypes'

const linebreakRegex = /\r\n|\n|\r/gm

function chunk<T = any[]>(arr: T[], chunkSize: number): T[][] {
  const R = []
  for (let i = 0, len = arr.length; i < len; i += chunkSize) R.push(arr.slice(i, i + chunkSize))
  return R
}

function transformCandidateValues(candidateRowsChunk: any[]): CandidateVotes {
  const name = candidateRowsChunk[0].childNodes[0].value
  const party = candidateRowsChunk[2].childNodes[0].value
  const voteValues = candidateRowsChunk[3].value
    .trim()
    .replace(/    /g, ' ')
    .replace(/  /g, ' ')
    .replace(/  /g, ' ')
    .replace(linebreakRegex, '')
    .split(' ')

  const [districtVotes, mailVotes, totalVotes, percentFromValid, percentFromAll] = voteValues

  return {
    name,
    party,
    districtVotes,
    mailVotes,
    totalVotes,
    percentFromValid,
    percentFromAll,
  }
}

export function parseDistrictResults(text: string): DistrictResults {
  const sanitizedText = text
    .trim()
    .replace(/\&nbsp/g, ' ')
    .replace(linebreakRegex, '')
  const districtHtml: DefaultTreeParentNode = parse5.parse(sanitizedText) as DefaultTreeParentNode
  // @ts-ignore
  const candidateRows = districtHtml.childNodes[0].childNodes[1].childNodes
  const headerRow = candidateRows.shift()
  const attrHeaders = headerRow.value.trim().replace(/   /g, '  ').replace(/   /g, '  ').split('  ')

  const candidateRowsChunksWithSummary = chunk(candidateRows, 4)
  const candidateRowsChunks = candidateRowsChunksWithSummary.slice(0, 15)
  const summaryRows = candidateRowsChunksWithSummary.slice(15, 18)
  const candidateValues = candidateRowsChunks.map(transformCandidateValues)

  return {
    attrHeaders,
    candidateValues,
  }
}
