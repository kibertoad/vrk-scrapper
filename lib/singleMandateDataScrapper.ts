import puppeteer, { Browser } from 'puppeteer'
import ltx from 'ltx'
import { findAllUrls } from './urlFinder'
import { parseDistrictResults } from './districtResultsParser'

const VRK_ROOT_URL = 'https://rezultatai.vrk.lt/'
const SINGLE_MANDATE_URL = '?srcUrl=/rinkimai/1104/1/1746/rezultatai/lt/rezultataiVienm.html'

const DISTRICT_TABLE_SELECTOR = 'table.partydata:nth-child(11)'

const DISTRICT_NAME_SELECTOR = '#pub-content > div:nth-child(2) > h3:nth-child(1)'
const DISTRICT_RESULT_TABLE_SELECTOR = 'table.partydata:nth-child(11)'

async function getDistrictUrls(browser: Browser): Promise<readonly string[]> {
  const page = await browser.newPage()
  await page.goto(`${VRK_ROOT_URL}${SINGLE_MANDATE_URL}`)
  await page.waitForSelector(DISTRICT_TABLE_SELECTOR)
  const districtTableHTML = await page.$eval(DISTRICT_TABLE_SELECTOR, (el) => {
    return el.innerHTML
  })
  const districtUrls = findAllUrls(districtTableHTML)
  return districtUrls
}

async function getDistrictResults(browser: Browser, districtUrl: string): Promise<any> {
  const page = await browser.newPage()
  await page.goto(`${VRK_ROOT_URL}${districtUrl}`)
  await page.waitForSelector(DISTRICT_RESULT_TABLE_SELECTOR)
  const resultsTableHTML = await page.$eval(DISTRICT_RESULT_TABLE_SELECTOR, (el) => {
    return el.innerHTML
  })
  const districtName = await page.$eval(DISTRICT_NAME_SELECTOR, (el) => {
    return el.innerHTML
  })

  const districtResults = parseDistrictResults(resultsTableHTML)

  return {
    districtName,
    districtResults,
  }
}

export async function scrapeSingleMandateData(): Promise<void> {
  const browser = await puppeteer.launch()

  try {
    const districtUrls = await getDistrictUrls(browser)
    const districtResults = await getDistrictResults(browser, districtUrls[0])
  } finally {
    await browser.close()
  }
}
