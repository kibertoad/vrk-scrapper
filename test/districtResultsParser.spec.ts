import fs from 'fs'
import path from 'path'
import { parseDistrictResults } from '../lib/districtResultsParser'

const tableHTML = fs
  .readFileSync(path.resolve(__dirname, 'resources/single_mandate_district_results_table.html'))
  .toString()

describe('districtResultsParser', () => {
  it('parses results correctly', () => {
    const result = parseDistrictResults(tableHTML)
    expect(result).toMatchSnapshot()
  })
})
