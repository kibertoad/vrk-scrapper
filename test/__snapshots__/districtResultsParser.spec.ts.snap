// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`districtResultsParser parses results correctly 1`] = `
Object {
  "attrHeaders": Array [
    "Kandidatas",
    "Iškėlė",
    "Paduotų balsų skaičius",
    "apylinkėse",
    "paštu",
    "iš viso",
    "% ;nuo galiojančių biuletenių",
    "% ;nuo dalyvavusių rinkėjų",
  ],
  "candidateValues": Array [
    Object {
      "name": "Mykolas MAJAUSKAS",
      "party": "Tėvynės sąjunga – Lietuvos krikščionys demokratai",
      "voteValues": Array [
        "4465",
        "1790",
        "6255",
        "29,71",
        "28,92",
      ],
    },
    Object {
      "name": "Viktorija ČMILYTĖ-NIELSEN",
      "party": "Lietuvos Respublikos liberalų sąjūdis",
      "voteValues": Array [
        "4582",
        "1467",
        "6049",
        "28,73",
        "27,97",
      ],
    },
    Object {
      "name": "Vytenis ŠKARNULIS",
      "party": "Laisvės partija",
      "voteValues": Array [
        "1543",
        "623",
        "2166",
        "10,29",
        "10,02",
      ],
    },
    Object {
      "name": "Darius KUOLYS",
      "party": "Politinė partija „Lietuvos sąrašas“",
      "voteValues": Array [
        "826",
        "217",
        "1043",
        "4,95",
        "4,82",
      ],
    },
    Object {
      "name": "Dovilė ŠAKALIENĖ",
      "party": "Lietuvos socialdemokratų partija",
      "voteValues": Array [
        "767",
        "254",
        "1021",
        "4,85",
        "4,72",
      ],
    },
    Object {
      "name": "Vitalija JANKAUSKAITĖ-MILČIUVIENĖ",
      "party": "Lietuvos valstiečių ir žaliųjų sąjunga",
      "voteValues": Array [
        "711",
        "239",
        "950",
        "4,51",
        "4,39",
      ],
    },
    Object {
      "name": "Kristina DZERŽINSKA",
      "party": "Lietuvos lenkų rinkimų akcija - Krikščioniškų šeimų sąjunga",
      "voteValues": Array [
        "526",
        "75",
        "601",
        "2,85",
        "2,78",
      ],
    },
    Object {
      "name": "Rasa IVANAUSKIENĖ",
      "party": "Darbo partija",
      "voteValues": Array [
        "461",
        "126",
        "587",
        "2,79",
        "2,71",
      ],
    },
    Object {
      "name": "Zenonas VAIGAUSKAS",
      "party": "Lietuvos socialdemokratų darbo partija",
      "voteValues": Array [
        "415",
        "118",
        "533",
        "2,53",
        "2,46",
      ],
    },
    Object {
      "name": "Robertas RAMANAUSKAS",
      "party": "Nacionalinis susivienijimas",
      "voteValues": Array [
        "409",
        "90",
        "499",
        "2,37",
        "2,31",
      ],
    },
    Object {
      "name": "Diana STOMIENĖ",
      "party": "Partija „Laisvė ir teisingumas“",
      "voteValues": Array [
        "265",
        "89",
        "354",
        "1,68",
        "1,64",
      ],
    },
    Object {
      "name": "Inga LAURUŠONĖ",
      "party": "Lietuvos žaliųjų partija",
      "voteValues": Array [
        "263",
        "63",
        "326",
        "1,55",
        "1,51",
      ],
    },
    Object {
      "name": "Sakalas GORODECKIS",
      "party": "Centro partija - tautininkai",
      "voteValues": Array [
        "255",
        "53",
        "308",
        "1,46",
        "1,42",
      ],
    },
    Object {
      "name": "Arvydas AKSTINAVIČIUS",
      "party": "Krikščionių sąjunga",
      "voteValues": Array [
        "165",
        "58",
        "223",
        "1,06",
        "1,03",
      ],
    },
    Object {
      "name": "Andrius NIKITINAS",
      "party": "Partija LIETUVA – VISŲ",
      "voteValues": Array [
        "119",
        "21",
        "140",
        "0,66",
        "0,65",
      ],
    },
  ],
}
`;
